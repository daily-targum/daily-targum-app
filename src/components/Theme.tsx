import React from 'react';
import { DefaultTheme, Provider as PaperProvider } from 'react-native-paper';
import { Appearance, AppearanceProvider } from 'react-native-appearance';
import { useSafeArea, EdgeInsets } from 'react-native-safe-area-context';
import { Provider as ContextThemeProvider, useTheme as defaultUseTheme, Context } from 'react-context-theming/lib/index';
import * as NativeTheming from 'react-context-theming/lib/native';
import { StyleCreatorFunction, GenerateStylesFunction } from 'react-context-theming/lib/native';
import { ComputedTheme } from '../types';
import { theme } from '../constants';
import { StyleSheet, Platform } from 'react-native';
import { useSelector } from '../store';

/**
 * THIS IS A HACK!!!
 * From some reason, edge inset top
 * always seems to be to large. It 
 * seems that multiplying it by 2/3 
 * gives a good approximation of the
 * actual height of the notch.
 */
function statusBarHeight(insets: EdgeInsets) {
  let {top} = insets;
  if(Platform.OS === 'ios') {
    top *= 2/3;
  }
  return Math.round(top);
}

export function Provider({ children }: {children: React.ReactNode}) {
  const insets = useSafeArea();
  const darkModeOverride = useSelector(s => s.theme.darkModeOverride);
  const useDeviceSettings = useSelector(s => s.theme.useDeviceSettings);
  const [ activeTheme, setActiveTheme ] = React.useState(darkModeOverride ? theme.dark : theme.light);
  const [ systemWideDarkMode, setSystemWideDarkMode ] = React.useState(false);

  function updateTheme() {
    const darkAppearance = Appearance.getColorScheme() === 'dark';
    const darkMode = useDeviceSettings ? darkAppearance : darkModeOverride;
    setActiveTheme(darkMode ? theme.dark : theme.light);
    setSystemWideDarkMode(darkAppearance);
  }

  // subscribe to dark mode
  React.useEffect(() => {
    if(useDeviceSettings) {
      const subscription = Appearance.addChangeListener(() => updateTheme());
      return function cleanup() {
        subscription.remove();
      };
    }
  }, [useDeviceSettings]);

  // subscribe to screen rotation
  React.useEffect(() => {
    updateTheme();
  }, [darkModeOverride, useDeviceSettings]);

  const {colors} = activeTheme;

  // pass theme props down
  // to React Native Paper
  const computedPaperTheme = {
    ...DefaultTheme,
    colors: {
      ...DefaultTheme.colors,
      primary: colors.primary,
      accent: colors.accent,
      text: colors.text,
      background: colors.background,
      surface: colors.surface,
      divider: colors.divider
    },
    dark: activeTheme.dark,
    roundness: activeTheme.roundness()
  };

  const computedTheme = {
    ...activeTheme, 
    insets, 
    systemWideDarkMode,
    statusBarHeight: statusBarHeight(insets)
  }

  return (
    <AppearanceProvider>
      <ContextThemeProvider<any> theme={computedTheme}>
        <PaperProvider theme={computedPaperTheme}>{children}</PaperProvider>
      </ContextThemeProvider>
    </AppearanceProvider>
  );
}

export const useTheme = (): ComputedTheme => defaultUseTheme<ComputedTheme>();

export function makeStyleCreator<
  T = ComputedTheme, 
  S extends StyleSheet.NamedStyles<S> | StyleSheet.NamedStyles<any> = never
>(
  styleFn: StyleCreatorFunction<T, S> | GenerateStylesFunction<T, S>
): GenerateStylesFunction<T, S> {
  return NativeTheming.makeStyleCreator<T, S>(styleFn);
}

export function useStyleCreator<
  T = ComputedTheme, 
  S = never
>(
  styleFn: GenerateStylesFunction<T, S>,
  ...extraData: any[]
) {
  return NativeTheming.useStyleCreator<T, S>(styleFn, ...extraData)
};

export function withStyleCreator<
  T = ComputedTheme,
  S = never
>(
  Component: any, 
  styleFn: GenerateStylesFunction<T, S>,
  ...extraData: any[]
): any {
  return NativeTheming.withStyleCreator<T, S>(Component, styleFn, ...extraData)
}

export function withTheme(Component: any): any {
  return class WrappedComponent extends React.Component<{}, null> {
    static contextType = Context;
    render() {
      return <Component {...this.props} theme={this.context}/>;
    }
  };
}

export default {
  Provider,
  useTheme,
  makeStyleCreator,
  useStyleCreator,
  withStyleCreator,
  withTheme
};