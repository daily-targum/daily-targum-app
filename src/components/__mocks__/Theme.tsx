import React from 'react';
import { theme } from '../../constants';
import { ComputedTheme } from '../../types';
import { makeStyleCreator, GenerateStylesFunction } from 'react-context-theming/lib/native';
import { Provider } from 'react-context-theming';

const THEME: ComputedTheme = {
  ...theme.light,
  insets: {
    ...theme.light.insets,
    top: 30
  },
  systemWideDarkMode: theme.light.dark,
  statusBarHeight: 30
};

export const useTheme: () => ComputedTheme = () => THEME;

export function useStyleCreator<
  S = never
>(
  styleFn: GenerateStylesFunction<ComputedTheme, S>,
  ...extraData: any[]
) {
  return styleFn(useTheme(), ...extraData);
};

export function withTheme(Component: any): any {
  return class WrappedComponent extends React.Component<{}, null> {
    render() {
      return <Component {...this.props} theme={THEME}/>;
    }
  };
}

export function withStyleCreator<
  S = never
>(
  Component: any, 
  styleFn: GenerateStylesFunction<ComputedTheme, S>,
  ...extraData: any[]
): any {
  return class WrappedComponent extends React.Component<{}, null> {
    render() {
      return (
        <Component 
          {...this.props} 
          styles={styleFn(THEME, ...extraData)}
          theme={this.context}
        />
      );
    }
  };
}

export default {
  Provider,
  useTheme,
  makeStyleCreator,
  useStyleCreator,
  withStyleCreator,
  withTheme
}
export {
  Provider,
  makeStyleCreator,
}