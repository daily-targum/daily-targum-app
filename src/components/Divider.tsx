import React from 'react';
import { View } from 'react-native';
import { useTheme } from './Theme';

function Divider({
  color
}: {
  color?: string
}) {
  const {colors} = useTheme();

  return (
    <View
      style={{
        height: 0.5,
        backgroundColor: color || colors.divider
      }}
    />
  );
}

export default Divider;