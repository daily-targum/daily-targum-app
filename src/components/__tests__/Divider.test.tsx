jest.mock('../Theme');

import React from 'react';
import Divider from '../Divider';
import { useTheme } from '../Theme';
import renderer from 'react-test-renderer';

describe('<Divider />', () => {

  it('renders correctly', () => {
    const tree = renderer.create(
      <Divider/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('inherits color from theme provider', () => {
    const tree: any = renderer.create(<Divider/>).toJSON();
    expect(tree?.props?.style[0]?.backgroundColor).toBe(useTheme().colors.divider);
  });

});