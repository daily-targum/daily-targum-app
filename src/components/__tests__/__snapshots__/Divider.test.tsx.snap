// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Divider /> renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0.1)",
      "height": 0.5,
    }
  }
/>
`;
